### PR TITLE
Allow MetricBuildItem to declare the desired registry type

### LINKED ISSUE #7061: Allow MetricBuildItem to declare the desired scope
Currently all metrics from `MetricBuildItem`s are placed into the `vendor` registry

### PR DESCRIPTION
**Pull request by jmartisk**
Fixes #7061

### CODE REVIEW

**Comment by mkouba** [comment]
Looks good but it's not related to a CDI scope @geoand ;-). So there's probably no special need for input from me or Matej...

**Comment by geoand** [comment]
LOL, my heuristic didn't work then :P

**Comment by jmartisk** [comment]
Whoops :)

**Review by gastaldi — APPROVED** [review]
LGTM, but IMHO I prefer builder methods without the `set` prefix :)

**Comment by manovotn** [comment]
Almost got me there, I was eyeing the code and thinking I might need an extra shot of caffeine to understand how is `MetricRegistry.Type.VENDOR` a CDI scope :-D

**Comment by geoand** [comment]
My ML algorithm needs to better training :P

**Comment by jmartisk** [comment]
Ok sounds like it will perhaps be a good idea to use the term "registry type" instead :D I will also remove the `set` prefix from the builder

File: extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricResource.java [hunk 1]
```diff
@@ -16,18 +16,34 @@
...
 
+    @Inject
+    @RegistryType(type = MetricRegistry.Type.BASE)
+    MetricRegistry baseRegistry;
+
     @GET
     @Path("/get-counters")
     @Produces("application/json")
-    public String[] getRegisteredCounters() {
+    public String[] getRegisteredCountersInVendorScope() {
```
Comment (gastaldi) [inline extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricResource.java]:
```suggestion
    public String[] getRegisteredCountersInVendorRegistryType() {
```

File: extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricResource.java [hunk 2]
```diff
@@ -16,18 +16,34 @@
...
         return vendorRegistry.getCounters().keySet().stream().map(MetricID::getName).toArray(String[]::new);
     }
 
+    @GET
+    @Path("/get-counters-base")
+    @Produces("application/json")
+    public String[] getRegisteredCountersInBaseScope() {
```
Comment (gastaldi) [inline extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricResource.java]:
```suggestion
    public String[] getRegisteredCountersInBaseRegistryType() {
```

File: extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionDisabledTestCase.java [hunk 3]
```diff
@@ -29,9 +31,17 @@
             .addBuildChainCustomizer(buildCustomizer());
 
     @Test
-    public void test() {
+    public void testBaseScope() {
```
Comment (gastaldi) [inline extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionDisabledTestCase.java]:
```suggestion
    public void testBaseRegistryType() {
```

File: extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionDisabledTestCase.java [hunk 4]
```diff
@@ -29,9 +31,17 @@
...
+        String[] metricNames = RestAssured.when().get("/get-counters-base").then().extract().as(String[].class);
+        assertThat(metricNames,
+                not(hasItemInArray("io.quarkus.smallrye.metrics.test.MetricResource.countMeInBaseScope")));
+        assertThat(metricNames, not(emptyArray())); // regular base metrics should still be present
+    }
+
+    @Test
+    public void testVendorScope() {
```
Comment (gastaldi) [inline extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionDisabledTestCase.java]:
```suggestion
    public void testVendorRegistryType() {
```

**Review by gastaldi — CHANGES_REQUESTED** [review]
Awesome! Added some comments with minor adjustments to the tests to match the new nomenclature ;)

**Comment by geoand** [comment]
@jmartisk when you are done with the fixes, can you please squash the commits? We don't want those `apply review suggestion` ones :P

File: extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionTestCase.java [hunk 5]
```diff
@@ -59,11 +76,17 @@ public void accept(BuildChainBuilder builder) {
...
-    public void test() {
+    public void testVendorScope() {
         String[] metricNames = RestAssured.when().get("/get-counters").then().extract().as(String[].class);
         assertThat(metricNames, Matchers.arrayContainingInAnyOrder(
                 "io.quarkus.smallrye.metrics.test.MetricResource.countMePlease",
                 "io.quarkus.smallrye.metrics.test.MetricResource.countMePlease2"));
     }
 
+    @Test
+    public void testBaseScope() {
```
Comment (gastaldi) [inline extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionTestCase.java]: Darn I missed these two tests :)

File: extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionTestCase.java [hunk 6]
```diff
@@ -22,8 +23,9 @@
 
 /**
  * Test the metric registration mechanism for other Quarkus extensions.
- * The test simulates an extension that registers a counter for every method
- * whose name starts with "countMePlease".
+ * The test simulates an extension that registers a counter in VENDOR scope for every method
```
Comment (gastaldi) [inline extensions/smallrye-metrics/deployment/src/test/java/io/quarkus/smallrye/metrics/test/MetricsFromExtensionTestCase.java]: scope is used here too

**Review by gastaldi — APPROVED** [review]
Perfection

**Pull request merged**